components/gitpod-db/src/typeorm/code-sync-resource-db.ts

```diff
@@ -0,0 +1,129 @@
+                const subQuery = qb.subQuery()
+                    .select('resource2.userId')
+                    .addSelect('resource2.kind')
+                    .addSelect('max(resource2.created)')
+                    .from(DBCodeSyncResource, 'resource2')
+                    .where('resource2.userId = :userId', { userId })
+                    .groupBy('resource2.kind')
+                    .orderBy('resource2.created', 'DESC')
+                    .getQuery();
+                return '(resource.userId,resource.kind,resource.created) IN ' + subQuery;
+            })
+            .getMany();
+
+        const latest: Record<ServerResource, string> = Object.create({});
+        const manifest: IUserDataManifest = { session: userId, latest };
+        for (const resource of resources) {
+            latest[resource.kind] = resource.rev;
+        }
+        return manifest;
+    }
+
+    async getResource(userId: string, kind: ServerResource, rev: string): Promise<DBCodeSyncResource | undefined> {
+        const connection = await this.typeORM.getConnection();
+        return this.doGetResource(connection.manager, userId, kind, rev);
+    }
+
+    async getResources(userId: string, kind: ServerResource): Promise<DBCodeSyncResource[]> {
+        const connection = await this.typeORM.getConnection();
+        return connection.manager.getRepository(DBCodeSyncResource)
+            .createQueryBuilder('resource')
+            .where('resource.userId = :userId AND resource.kind = :kind', { userId, kind })
+            .orderBy('resource.created', 'DESC')
+            .getMany();
+    }
+
+    async delete(userId: string): Promise<void> {
+        const connection = await this.typeORM.getConnection();
+        await connection.createQueryBuilder()
+            .delete()
+            .from(DBCodeSyncResource)
+            .where("userId = :userId", { userId })
+            .execute();
+    }
+
+    async insert(userId: string, kind: ServerResource, doInsert: (rev: string, oldRev?: string) => Promise<void>, params?: CodeSyncInsertOptions): Promise<string | undefined> {
+        const connection = await this.typeORM.getConnection();
+        return await connection.transaction(async manager => {
+            let latest: DBCodeSyncResource | undefined;
+            let toUpdated: DBCodeSyncResource | undefined;
+            if (params?.revLimit) {
+                const resources = await this.doGetResources(manager, userId, kind);
+                latest = resources[0];
+                if (resources.length >= params.revLimit) {
+                    toUpdated = resources[resources.length - 1];
+                }
+            } else {
+                latest = await this.doGetResource(manager, userId, kind, 'latest');
+            }
+            if (params?.latestRev && latest?.rev !== params.latestRev) {
+                return undefined;
+            }
+            const rev = uuid.v4();
+            await doInsert(rev, toUpdated?.rev);
```

Comment:
what effect do the long-running operations within the transaction have? Do we lock tables in the meantime?

Reply:
good point 👍 It would be good if it only blocks a user partition of the table but not for other users.

Reply:
Yes, we lock the table but only rows which are affected by looking up the latest resource, i.e. only rows belonging to a user. I think it is what we need.